### ISSUE TITLE
Support defining NDArray-like types for NumPy-ish frameworks?

### ISSUE THREAD

**Issue by EricCousineau-TRI** [ramonhagenaars/nptyping#30]
I'd like to define something like `Tensor` for use with `pytorch` (relates #27).

Some random notes:

- This would loosely enable any frameworks that offer array-like objects (e.g. `torch`, `tensorflow`, `open3d` ([code](https://github.com/intel-isl/Open3D/blob/970d9af7e679de14eb0cff1595f932c36463bad1/cpp/pybind/core/tensor.cpp))), generally something that offers `.shape` and `.dtype` that behave similar to NumPy's
- Ideally, some of this work could also permit ndarray subclasses ([docs](https://numpy.org/doc/stable/user/basics.subclassing.html)), and handle the whole covariance shindig mentioned in PEP 484 ([docs](https://www.python.org/dev/peps/pep-0484/#covariance-and-contravariance))

**Comment by EricCousineau-TRI** [comment]
I, uh, couldn't figure out how to make PyTorch easily play well with `_NDArray` and all the generics, so I went ahead and wrote a poor-man's duplicate. I've hashed out a simple version of this here:
https://github.com/EricCousineau-TRI/repro/blob/4a5c71162bf2e3c0afc4d08c5d757aa0841000c4/python/nptyping_ish/test/typing_test.py

**Comment by EricCousineau-TRI** [comment]
FWIW I've updated the prototype to allow working with `dataclasses` and as a (semi-annoying) stand-in for `default_collate`:
https://github.com/EricCousineau-TRI/repro/blob/7a08250e55899dd47fb77ac6524331cd24d86e59/python/nptyping_ish/typing_/test/typing_test.py#L293-L427

**Comment by ramonhagenaars** [comment]
Good idea to support similar frameworks. I was thinking of adding support for `pandas`, `tensorflow` and `pytorch` would be nice additions too.

The other frameworks are to become optional for `nptyping`, meaning that you have to explicitly specify them when installing:

```
pip install nptyping[pandas]
pip install nptyping[tensorflow]
...
pip install nptyping[complete]
```

I think I'll start designing an addition for Pandas first. Then I could use help with other frameworks (I don't know `open3d` for example).

**Comment by EricCousineau-TRI** [comment]
Sounds great!

And yeah, for `open3d`, I think it'd basically be a restricted subset of `NumPy` functionality, so nothing fancy. (Just may have `device`-like semantics like PyTorch / TensorFlow, but no computation graph thingers for reverse autodiff going on.)

**Comment by zplizzi** [comment]
Would love to see this for torch :)

**Comment by dxm447** [comment]
Would love to see it for cupy too

**Comment by dxm447** [comment]
`cupy.typing` has been available since [this year](https://github.com/cupy/cupy/pull/6251), and is a like for like replacement of `np.typing`
